src/drivers/intel/cavs/hda-dma.c

```diff
@@ -442,22 +385,45 @@ static int hda_dma_host_copy(struct dma *dma, int channel, int bytes,
 {
 	struct dma_pdata *p = dma_get_drvdata(dma);
 	struct hda_chan_data *chan = p->chan + channel;
+	uint64_t deadline;
+
+	tracev_hddma("hda-dmac: %d channel %d -> copy 0x%x bytes",
+		     dma->plat_data.id, chan->index, bytes);
 
 	if (channel >= HDA_DMA_MAX_CHANS) {
 		trace_hddma_error("hda-dmac: %d invalid channel %d",
 				  dma->plat_data.id, channel);
 		return -EINVAL;
 	}
 
-	if (flags & DMA_COPY_PRELOAD)
-		chan->state |= HDA_STATE_HOST_PRELOAD;
+	hda_dma_get_dbg_vals(chan, HDA_DBG_PRE, HDA_DBG_HOST);
 
-	if (chan->state & HDA_STATE_INIT)
-		return 0;
-	else if (chan->state & HDA_STATE_HOST_PRELOAD)
-		return hda_dma_host_preload(dma, chan);
-	else
-		return hda_dma_host_copy_ch(dma, chan, bytes);
+	if (!(flags & DMA_COPY_PRELOAD))
+		/* set BFPI to let host gateway know we have read size,
+		 * which will trigger next copy start.
+		 */
+		hda_dma_inc_fp(dma, chan->index, bytes);
+
+	/* blocking mode copy */
+	if (flags & DMA_COPY_BLOCKING) {
+		deadline = platform_timer_get(platform_timer) +
+			clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) *
+			PLATFORM_HOST_DMA_TIMEOUT / 1000;
+
+		while (hda_dma_get_copied_bytes(dma, chan) < bytes) {
```

Comment:
and then here avail < bytes looks something not accurate?
e.g. for playback, we have half buffer available already, then schedule a new copy, bytes = half buffer, and then here it will be true at once?

Reply:
@keyonjie You cannot be in such situation, because playback always needs to be preloaded (buffer full), so after every pointer increment DMA buffer will be filled again. The only situation where we could exit too early for playback or capture is when DMA stops working and transferring data. But still, you won't be able to execute the next dma copy due to not enough avail/free space, so no harm done.

Reply:
@tlauda no, I think here it will exit too early ALWAYS for the ping-pong buffer playback.
1. preload --> buffer full.
2. one period consumed, call hda_dma_inc_fp(), w_ptr = 0, r_ptr = half full.
3. then go to this check: while (hda_dma_get_copied_bytes(dma, chan) < bytes)
Here the copied_bytes will return at least half full, even the DMA new copying is not started, so this check actually will be always true, it is not functional.